File: strings/add_and_search_words_data_structure.rb
class TrieNode
    attr_accessor :children, :end_of_word

    def initialize()
        @children = {}
        @end_of_word = false
    end
end

class WordDictionary
    attr_reader :root

    def initialize()
        @root = TrieNode.new
    end

=begin
    :type word: String
    :rtype: Void
=end
    def add_word(word)
        curr = root

        word.each_char do |ch|
            if !curr.children.key?(ch)
                curr.children[ch] = TrieNode.new
            end
            curr = curr.children[ch]
        end
        curr.end_of_word = true
    end

=begin
    :type word: String
    :rtype: Boolean
=end
    def search(word)
       @word = word
       dfs(0, root) 
    end

    def dfs(j, node)
        curr = node
        (j...@word.size).each do |i|
            ch = @word[i]
            if ch == '.'
                curr.children.values.each do |child|
                    return true if dfs(i + 1, child)
                end
                return false
            else
                return false if !curr.children.key?(ch)

                curr = curr.children[ch]
            end
        end
        curr.end_of_word
    end
end

# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary.new()
# obj.add_word(word)
# param_2 = obj.search(word)

File: strings/add_and_search_words_data_structure.py
class TrieNode:
    def __init__(self, val):
        self.val = val
        self.children = {}
        self.isEnd = False

class WordDictionary:

    def __init__(self):
        self.root = TrieNode("")

    def addWord(self, word: str) -> None:
        root = self.root
        for i in range(len(word)):
            if word[i] not in root.children:
                root.children[word[i]] = TrieNode(word[i])    
            if i == len(word) - 1:
                root.children[word[i]].isEnd = True
            root = root.children[word[i]]
        root.isEnd = True
                

    def search(self, word: str) -> bool:
        def dfs(node, i):
            if i == len(word):
                return node.isEnd

            if word[i] == '.':
                for child in node.children.values():
                    if dfs(child, i + 1):
                        return True
                return False

            if word[i] not in node.children:
                return False

            return dfs(node.children[word[i]], i + 1)

        return dfs(self.root, 0)

# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)